[ios][android][background-task] add new package: expo-background-task

chrfalch opened the pull request:
# Why

`expo-background-fetch` uses deprecated APIs and we'd like to replace these on both iOS and Android using the latest and greatest APIs. With this new package we can deprecate the `expo-background-fetch` package and move on.

The new APIs are [`BGTaskScheduler`](https://devhttps://developer.apple.com/documentation/backgroundtasks/bgtaskscheduler) API on iOS and the [`WorkManager`](https://developer.android.com/topic/libraries/architecture/workmanager) API on Android.

These are the recommended APIs for background tasks.

# How

- Added expo-background-task package
- Added documentation for the background-task package
- Added test screen in BareExpo
- Added necessary UIBackgroundModes in Expo-go and Bare-Expo

# Test Plan

- Run in BareExpo on device (iOS) or simulator (android) and open the BackgroundTask screen
- Test registration/unregistration
- Test running tasks
  - Test running tasks only runs the `background-task` (see log output)
  - Ensure that tasks are periodically repeated

# Checklist

- [X] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [X] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [X] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

chrfalch committed:
[expo-background-task] Added new package to bare-expo

chrfalch committed:
[expo-background-task] Added new package

chrfalch committed:
Updated implementation (ts/android) to be aligned with background-fetch

- Added optional import in app.tsx (BareExpo)
- Simplified BackgroundTaskScreen
- Implemented Kotlin API
- Updated TS build
- Added callback parameter to TaskManagerUtils executeTask

chrfalch committed:
WIP: Working on detecting if we're in startup when calling executeTask

chrfalch committed:
Android background-task module

- Simplified and added some utility functions for Operations and Futures
- Adjustments to the implementation

chrfalch committed:
Cleared up some comments

chrfalch committed:
Fixed mismatch after prev commit

chrfalch committed:
[ios] Added EXTaskLaunchReasonBackgroundTask to launch reasons

To be able to run tasks through the BGTaskScheduler class on iOS we need to be able to add a new item to the `EXTaskLaunchReason` structure in the TaskManager interfaces so that we can run tasks that should be run by the background task package.

chrfalch committed:
[ios] Added background task result type to be able to use with iOS expected results

As a legacy from background fetch on iOS each task needs to return a value indicating the results of the task. This was tightly connected to the BackgroundFetch result values - in this commit I've included the conversion from the new background-task package to these old values.

chrfalch committed:
Example is now using the new return value types

chrfalch committed:
[ios] added ios implementation of background tasks

- Uses BGTaskScheduler
- Executes registered tasks

WIP: No handling of intervals yet

chrfalch committed:
Ran swiftlint

chrfalch committed:
moved scheduler's single method to service

chrfalch committed:
Moved service -> scheduler

chrfalch committed:
added example to comment/docs

chrfalch committed:
added error handling to screen in ncl

chrfalch committed:
Added comment

chrfalch committed:
Added typing for parameters

chrfalch committed:
[android] Added handling of miminum interval

chrfalch committed:
Adjusted background task example screen

Changed some variable names and cleaned up some white space

chrfalch committed:
Cleaned up some comments/types after reviewing docs

chrfalch committed:
Cleaned up example screen

chrfalch committed:
Check for background + updated options, docs and types

Minimum interval is now one day on Android
Android waits untill minimum interval before starting in Release mode
Android skips running if the app is in the foreground.

chrfalch committed:
Added function for trigger in debug mode on iOS

chrfalch committed:
Made sure we're only running registered tasks when there are multiple tasks registered

chrfalch committed:
Linting

chrfalch committed:
Added documentation

- Added documentation for expo-background-task
- Added reference from task-manager
- Added api docs for background-tasks
- Included background-task module in GenerateDocsAPIData tool

chrfalch committed:
Added processing to UIBackgroundModes

chrfalch committed:
Updated link to the BGTaskScheduler API docs

chrfalch committed:
swift lint

chrfalch committed:
Added changelog entries

chrfalch committed:
lint docs

chrfalch committed:
Added changelog entry to expo-task-manager

chrfalch committed:
Removed unused build file

chrfalch committed:
yarn build plugin

chrfalch committed:
android/spotless

brentvatne commented:
@lukmccall @alanjhughes @Kudo - feel free to look at this if you have some time, and if you would like, @chrfalch would be happy to walk you through it on a call!

dhirajksharma commented:
I have been trying to use `expo-background-fetch` for setting up some background tasks in my expo managed project, but they weren't getting executed. Then I came across this [issue](https://github.com/expo/expo/issues/33404) according to which `expo-doctor` says that `expo-background-fetch` is unsupported on the new architecture. 

I ran `expo-doctor` in my project and it said the same thing about the package. Is it that the use of deprecated APIs in `expo-background-fetch` is causing it's inability to work on the new architecture?

lukmccall commented on packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskScheduler.kt:
```diff
@@ -0,0 +1,214 @@
+     */
+    suspend fun Operation.await(): Boolean = suspendCancellableCoroutine { continuation ->
+      val future = this.result
+      val executor = Executors.newCachedThreadPool()
```
I'm not sure if creating a thread pool for each operation is a good idea, but I don't know how we can change it.

chrfalch replied:
I made the thread pool a companion object on the class - so that it doesn't create a new one each time the function is called.

Nice catch!

chrfalch committed:
Update packages/expo-background-task/android/src/main/AndroidManifest.xml

Co-authored-by: Łukasz Kosmaty <lukasz.kosmaty@swmansion.com>

chrfalch commented:
> I have been trying to use `expo-background-fetch` for setting up some background tasks in my expo managed project, but they weren't getting executed. Then I came across this [issue](https://github.com/expo/expo/issues/33404) according to which `expo-doctor` says that `expo-background-fetch` is unsupported on the new architecture.

Hi, could you please create a small issue about this? Expo-background-fetch should work with the new architecture.

chrfalch committed:
codereview: renamed variables, removed java-prefix m

chrfalch committed:
codereview: moved variables outside class

chrfalch committed:
codereview: simplified const declarations

brentvatne committed:
[docs] Adjust based on reading through

brentvatne commented on docs/pages/versions/unversioned/sdk/background-task.mdx:
```diff
@@ -0,0 +1,208 @@
+
+### When will they stop?
+
+The native platform APIs will stop running background tasks if the user explicitly kills the app (they will be resumed when the user restarts the app) or if the app is uninstalled. If the device is rebooted or the app crashes, the background tasks will be resumed.
```
just wanted to double check - is this accurate? there isn't any way to run a task when the app has been killed?

chrfalch replied:
As discussed in our sync discussion I'll add a bit more info about how this works on Android.

brentvatne committed:
[docs] Update for consistency with background task

brentvatne commented:
feedback in: https://github.com/expo/expo/pull/33438/commits/2e056ead9ef34d2fdf8857a8017418b2fb0452d4

related: https://bsky.app/profile/notbrent.dev/post/3lcyl2ibbfk2k 😅 

feel free to revert anything that is incorrect

chrfalch committed:
codereview: Remove unecessary launch - already in scope

lukmccall commented on docs/pages/versions/unversioned/sdk/background-task.mdx:
```diff
@@ -0,0 +1,208 @@
+import { Collapsible } from '~/ui/components/Collapsible';
+import { PlatformTags } from '~/ui/components/Tag';
+
+`expo-background-task` provides an API to run deferrable background tasks in a way that optimises battery and power consumption for the end user. This module uses the [`BGTaskScheduler`](https://developer.apple.com/documentation/backgroundtasks/bgtaskscheduler) API on iOS and the [`WorkManager`](https://developer.android.com/topic/libraries/architecture/workmanager) API on Android to schedule tas ...
+
```
```suggestion
`expo-background-task` provides an API to run deferrable background tasks in a way that optimizes battery and power consumption for the end user. This module uses the [`BGTaskScheduler`](https://developer.apple.com/documentation/backgroundtasks/bgtaskscheduler) API on iOS and the [`WorkManager`](https://developer.android.com/topic/libraries/architecture/workmanager) API on Android to schedule tasks. It also uses the [expo-task-manager](task-manager.mdx) Native API for running JavaScript tasks.

```

lukmccall commented on docs/pages/versions/unversioned/sdk/background-task.mdx:
```diff
@@ -0,0 +1,208 @@
+
+## Background tasks
+
+A background task is a deferrable unit of work that is performed in the background, outside the lifecycle of your app. This is useful for tasks that need to be executed when the app is not active, such as syncing data with a server, fetching new content, or even checking if there are any [Expo Updates](updates.mdx).
```
```suggestion
A background task is a deferrable unit of work that is performed in the background, outside the lifecycle of your app. This is useful for tasks that need to be executed when the app is inactive, such as syncing data with a server, fetching new content, or even checking if there are any [Expo Updates](updates.mdx).
```

lukmccall commented on packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskConsumer.kt:
```diff
@@ -0,0 +1,116 @@
+
+  private var task: TaskInterface? = null
+  private var isBackgrounded: Boolean = true
+  private val taskCoroutineScope = CoroutineScope(Dispatchers.Default)
```
We probably need to destroy this scope in `onHostDestroy` (or somewhere else - I'm not sure if the consumer can outlive the rn host) 🤔

lukmccall commented on packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskConsumer.kt:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  private fun getIntervalMinutes(): Long {
+    val options = if (task != null) task!!.options else null
+
+    if (options != null && options.containsKey("minimumInterval")) {
+      // minimumInterval option is in minutes
+      return (options["minimumInterval"] as Number).toLong()
+    }
+    return DEFAULT_INTERVAL_MINUTES
```
```suggestion
		return (task?.options?.["minimumInterval"] as? Number)?.toLong() 
		  ?: DEFAULT_INTERVAL_MINUTES
```

chrfalch replied:
Had to go with this one to avoid syntax errors:

```kotlin
val options = task?.options as? Map<String, Any?>
    return (options?.get("minimumInterval") as? Number)?.toLong()
      ?: DEFAULT_INTERVAL_MINUTES
```

lukmccall commented on packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskScheduler.kt:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    /**
+     Returns true if the worker task is pending
```
```suggestion
     * Returns true if the worker task is pending
```

lukmccall commented on packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskScheduler.kt:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    /**
+     Cancels the worker task
```
```suggestion
     * Cancels the worker task
```

lukmccall commented on packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskScheduler.kt:
```diff
@@ -0,0 +1,214 @@
+    private val TAG: String = BackgroundTaskScheduler::class.java.simpleName
+
+    /**
+     Schedules the worker task to run. The worker should run periodically.
```
```suggestion
     * Schedules the worker task to run. The worker should run periodically.
```

lukmccall commented on packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskScheduler.kt:
```diff
@@ -0,0 +1,214 @@
+      // Ensure we have the react context
+      Log.i(TAG, "Enqueuing worker with identifier $WORKER_IDENTIFIER")
+
+      val data = Data.Builder()
+      data.putString("appScopeKey", appScopeKey)
+
+      // Create the work request
+      val builder = PeriodicWorkRequestBuilder<BackgroundTaskWork>(
+        repeatIntervalTimeUnit = TimeUnit.MINUTES,
+        repeatInterval = intervalMinutes
+      ).setInputData(data.build())
+        .setConstraints(
+          Constraints.Builder()
+            .setRequiresBatteryNotLow(true)
+            .setRequiredNetworkType(NetworkType.CONNECTED)
+            .build()
+        )
```
```suggestion
      val data = Data.Builder()
        .putString("appScopeKey", appScopeKey)
        .build()

			val constraints = Constraints.Builder()
      	.setRequiresBatteryNotLow(true)
        .setRequiredNetworkType(NetworkType.CONNECTED)
        .build()
      
      // Create the work request
      val builder = PeriodicWorkRequestBuilder<BackgroundTaskWork>(
        repeatIntervalTimeUnit = TimeUnit.MINUTES,
        repeatInterval = intervalMinutes
      )
      	.setInputData(data)
        .setConstraints(constraints)
```

lukmccall commented on packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskScheduler.kt:
```diff
@@ -0,0 +1,214 @@
+      Log.i(TAG, "runTasks: number of consumers ${consumers.size}")
+
+      val tasks = consumers.mapNotNull { consumer ->
+        if (consumer.taskType() == BackgroundTaskConsumer.BACKGROUND_TASK_TYPE) {
```
Instead of checking `taskType`, can't we just try to cast it to `BackgroundTaskConsumer`?

chrfalch commented on packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskScheduler.kt:
```diff
@@ -0,0 +1,214 @@
+
+      future.addListener({
+        try {
+          future.run { get() }
```
I tried this, but got the following warning when changing to `future.get()`

`Possibly blocking call in non-blocking context could lead to thread starvation `

chrfalch committed:
[codereview] Added adjustments from cr

chrfalch committed:
[codereview] Added throwing when calling test method in prod-build

chrfalch commented on packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskConsumer.kt:
```diff
@@ -0,0 +1,119 @@
+      Log.i(TAG, "Executing task '${mTask?.name}'")
+      taskManagerUtils.executeTask(mTask, null, callback)
+    } else {
+      Log.w(TAG, "Task was not executed since the app was not in the background or in Debug mode.")
```
The idea is to try to avoid calling the task when in the background, since it is asynchronous we might get into situations where a task is scheduled when in the background - but run when in the foreground. It should not fail, just warn. But maybe it is better to just skip logging - I think I added the log to test when developing. :) :)

chrfalch commented on packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskScheduler.kt:
```diff
@@ -0,0 +1,214 @@
+        )
+
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+        // Add minimum interval here as well so that the work doesn't start immediately
+        // In Debug builds we wait 30 seconds to make sure background tasks are easy to test.
+        // Remember to put the app in the background for the task to run.
+        if (ReactBuildConfig.DEBUG) {
+          builder.setInitialDelay(Duration.ofSeconds(30))
+        } else {
+          builder.setInitialDelay(Duration.ofMinutes(intervalMinutes))
+        }
```
I wrote a section about testing in the documentation that explains this - the idea was that you can add a test button in your dev builds to be able to test background tasks directly - it is cumbersome to wait for testing.

Kudo replied:
don't like the hard coded value here. not 100% if that's feasible, maybe to use an android BroadcastReceiver to receive the configuration only on debug build. then people can use adb to configurate the delay time, e.g.
`adb shell am broadcast -a com.example.app.DEBUG_ACTION --es config_key config_value`

chrfalch committed:
Removed logging when trying to execute while not in the background

chrfalch committed:
Update docs/pages/versions/unversioned/sdk/background-task.mdx

Co-authored-by: Łukasz Kosmaty <lukasz.kosmaty@swmansion.com>

chrfalch committed:
Update docs/pages/versions/unversioned/sdk/background-task.mdx

Co-authored-by: Łukasz Kosmaty <lukasz.kosmaty@swmansion.com>

chrfalch committed:
fixed import

chrfalch committed:
[codereview] added cancelling of scope

chrfalch committed:
[ci] fixed android/lint (ran ./gradlew spotlessApply)

Watchara242540 commented:
รำคาญโครตๆๆๆ

ในวันที่ พฤ. 12 ธ.ค. 2024 17:03 น. Alan Hughes ***@***.***>
เขียนว่า:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskConsumer.kt
> <https://github.com/expo/expo/pull/33438#discussion_r1881699865>:
>
> > +  override fun onHostDestroy() {
> +    taskCoroutineScope.cancel()
> +  }
>
> ⬇️ Suggested change
>
> -  override fun onHostDestroy() {
> -    taskCoroutineScope.cancel()
> -  }
> +  override fun onHostDestroy() {
> +      try {
> +        taskCoroutineScope.cancel(ModuleDestroyedException())
> +      } catch (e: IllegalStateException) {
> +        Log.e(TAG, "The scope does not have a job in it")
> +      }
> +  }
>
> This can throw is the scope does not contain a job
> ------------------------------
>
> In
> packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskScheduler.kt
> <https://github.com/expo/expo/pull/33438#discussion_r1881718005>:
>
> > +      // Get work manager
> +      val workManager = WorkManager.getInstance(context)
> +
> +      return try {
> +        val workInfos = workManager.getWorkInfosForUniqueWork(WORKER_IDENTIFIER).await()
> +        return workInfos.firstOrNull()
> +      } catch (e: Exception) {
> +        Log.i(TAG, "Calling getWorkInfosForUniqueWork failed with error " + e.message)
> +        null
> +      }
> +    }
> +
> +    /**
> +     * Helper function for calling functions returning an Operation
> +     */
> +    suspend fun Operation.await(): Boolean = suspendCancellableCoroutine { continuation ->
>
> ⬇️ Suggested change
>
> -    suspend fun Operation.await(): Boolean = suspendCancellableCoroutine { continuation ->
> +suspend fun Operation.await(): Boolean = withContext(Dispatchers.IO) {
> +  try {
> +    result.get()
> +    true
> +  } catch (e: Exception) {
> +    throw e
> +  }
> +}
>
> ------------------------------
>
> In
> packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskScheduler.kt
> <https://github.com/expo/expo/pull/33438#discussion_r1881723609>:
>
> > +        if (ReactBuildConfig.DEBUG) {
> +          builder.setInitialDelay(Duration.ofSeconds(30))
> +        } else {
> +          builder.setInitialDelay(Duration.ofMinutes(intervalMinutes))
> +        }
> +      }
> +
> +      // Create work request
> +      val workRequest = builder.build()
> +
> +      // Get Work manager
> +      val workManager = WorkManager.getInstance(context)
> +
> +      // Enqueue the work
> +      return try {
> +        val operation = workManager.enqueueUniquePeriodicWork(
>
> Is the result unused?
> ------------------------------
>
> In
> packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskScheduler.kt
> <https://github.com/expo/expo/pull/33438#discussion_r1881731914>:
>
> > +    suspend fun <T> ListenableFuture<T>.await(): T = suspendCancellableCoroutine { continuation ->
> +      val executor = Executors.newCachedThreadPool()
> +
> +      this.addListener({
> +        try {
> +          val result = runCatching { this.get() }
> +          continuation.resume(result.getOrThrow())
> +        } catch (e: Exception) {
> +          continuation.resumeWithException(e)
> +        } finally {
> +          executor.shutdown()
> +        }
> +      }, executor)
> +
> +      continuation.invokeOnCancellation {
> +        this.cancel(true)
> +        executor.shutdown()
> +      }
> +    }
> +  }
>
> ⬇️ Suggested change
>
> -    suspend fun <T> ListenableFuture<T>.await(): T = suspendCancellableCoroutine { continuation ->
> -      val executor = Executors.newCachedThreadPool()
> -
> -      this.addListener({
> -        try {
> -          val result = runCatching { this.get() }
> -          continuation.resume(result.getOrThrow())
> -        } catch (e: Exception) {
> -          continuation.resumeWithException(e)
> -        } finally {
> -          executor.shutdown()
> -        }
> -      }, executor)
> -
> -      continuation.invokeOnCancellation {
> -        this.cancel(true)
> -        executor.shutdown()
> -      }
> -    }
> -  }
> +suspend fun <T> ListenableFuture<T>.await(): T = withContext(Dispatchers.IO) {
> +  try {
> +    get()
> +  } catch (e: CancellationException) {
> +    cancel(true)
> +    throw e
> +  } catch (e: Exception) {
> +    throw e
> +  }
> +}
>
> Using the IO Dispatcher means we can drop the executor
> ------------------------------
>
> In
> packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskScheduler.kt
> <https://github.com/expo/expo/pull/33438#discussion_r1881736858>:
>
> > +import com.facebook.react.common.build.ReactBuildConfig
> +import com.google.common.util.concurrent.ListenableFuture
> +import expo.modules.interfaces.taskManager.TaskServiceProviderHelper
> +import kotlinx.coroutines.CompletableDeferred
> +import kotlinx.coroutines.awaitAll
> +import kotlinx.coroutines.suspendCancellableCoroutine
> +import java.time.Duration
> +import java.util.concurrent.Executors
> +import java.util.concurrent.TimeUnit
> +import kotlin.coroutines.resume
> +import kotlin.coroutines.resumeWithException
> +
> +class BackgroundTaskScheduler {
> +  companion object {
> +    // Shared executor
> +    val sharedExecutor = Executors.newCachedThreadPool()
>
> We're using coroutines so we can drop the executor
> ------------------------------
>
> In packages/expo-background-task/ios/BackgroundTaskScheduler.swift
> <https://github.com/expo/expo/pull/33438#discussion_r1881746322>:
>
> > @@ -0,0 +1,80 @@
> +// Copyright 2024-present 650 Industries. All rights reserved.
> +import BackgroundTasks
> +
> ***@***.***(BackgroundTaskScheduler)
>
> ⬇️ Suggested change
>
> ***@***.***(BackgroundTaskScheduler)
> ***@***.***
>
> The names are the same so we don't need to specify it
> ------------------------------
>
> In packages/expo-background-task/ios/BackgroundTaskScheduler.swift
> <https://github.com/expo/expo/pull/33438#discussion_r1881749172>:
>
> > +      @unknown default:
> +        print("An unknown BGTaskScheduler error occurred.")
> +        // Handle any future cases added by Apple
> +      }
> +    } catch {
> +      // All other errors
> +      throw CouldNotRegisterWorkerTask("Unknown error occurred.")
> +    }
> +  }
> +
> +  /**
> +   Calls a private iOS API if we're in debug mode to invoke the BGTaskScheduler's worker.
> +   NOTE: This code is only compiled when we're in DEBUG mode!
> +   */
> +  public static func triggerTaskForTesting() {
> +#if DEBUG && !targetEnvironment(simulator)
>
> In objc, this is omitted during preproceesing so won't be compiled. In
> Swift, this is evaluated at compile time, so this code will still end up in
> the binary and caught by app review. We may need to obfuscate it somehow
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/expo/expo/pull/33438#pullrequestreview-2498579503>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BJCBR44HLIXYO65UPS6GMA32FFNPJAVCNFSM6AAAAABTAHHFX2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDIOJYGU3TSNJQGM>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

chrfalch committed:
Update packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskConsumer.kt

Co-authored-by: Alan Hughes <30924086+alanjhughes@users.noreply.github.com>

chrfalch committed:
[codereview] applied comments from Alan

chrfalch committed:
[chore] spotless/lint

chrfalch committed:
[codereview] Removed unused result from enqueueUniquePeriodicWork

chrfalch committed:
Update packages/expo-background-task/ios/BackgroundTaskScheduler.swift

Co-authored-by: Alan Hughes <30924086+alanjhughes@users.noreply.github.com>

chrfalch committed:
[codereview] moved #ifdef code to obj-c

- Moved the test code to obj-c to ensure it is not included when building
- Added rescheduling of tasks after excution
- Added disabling of test button in BareExpo screen when task is unregistered.

chrfalch commented on packages/expo-background-task/ios/BackgroundTaskScheduler.swift:
```diff
@@ -0,0 +1,80 @@
+   NOTE: This code is only compiled when we're in DEBUG mode!
+   */
+  public static func triggerTaskForTesting() {
+#if DEBUG && !targetEnvironment(simulator)
```
Resolved by moving this to objective-c as discussed

Kudo commented on packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskConsumer.kt:
```diff
@@ -0,0 +1,116 @@
+   * This only applies when the app is in release mode.
+   */
+  fun executeTask(callback: TaskExecutionCallback) {
+    if (isBackgrounded || ReactBuildConfig.DEBUG) {
```
i'm not quite clear for this. the comment said "This only applies when the app is in release mode." but the code is checking for debug build. could you help to explain more?

chrfalch replied:
I will refactor this a bit - I agree with you.

Kudo commented on packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskConsumer.kt:
```diff
@@ -0,0 +1,116 @@
+      // Check if we have a task that
+      val tasks = taskService.getTasksForAppScopeKey(appScopeKey)
+      val ourTasks = tasks.filter { it.getString("taskType") == BACKGROUND_TASK_TYPE }
+      if (ourTasks.isEmpty()) {
```
note that `Dispatchers.Default` is a [thread pool](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-dispatchers/-default.html). if you intend to execute this code in concurrent mode, maybe you should add a lock here. otherwise, maybe try to use other coroutine with single thread.

Kudo commented on packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskExceptions.kt:
```diff
@@ -0,0 +1,13 @@
+
+class TaskMangerInterfaceNotFoundException : CodedException(message = "TaskManagerInterface not found")
+
+class TestMethodNotAvailableInProductionBuild : CodedException(message = "Background tasks cannot be triggered in production builds")
```
these class can be "internal"

Kudo commented on packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskModule.kt:
```diff
@@ -0,0 +1,47 @@
+import expo.modules.kotlin.modules.ModuleDefinition
+
+class BackgroundTaskModule : Module() {
+  companion object {
+    private const val MODULE_NAME: String = "ExpoBackgroundTask"
+    private val TAG: String = BackgroundTaskModule::class.java.simpleName
+  }
```
to clarify, do you mean only `MODULE_NAME` or also `TAG`? are you leaning toward to write TAG in private const val or in companion object?

Kudo commented on packages/expo-background-task/ios/BackgroundTaskAppDelegate.h:
```diff
@@ -0,0 +1,13 @@
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface BackgroundTaskAppDelegate : EXSingletonModule <UIApplicationDelegate>
```
the EXSingletonModule is mostly deprecated. could you share the intention to use here? if the idea is to handle `application:didFinishLaunchingWithOptions:`, you could use an [AppDelegate Subscriber](https://docs.expo.dev/modules/appdelegate-subscribers/#example). then we can also use swift than objc

chrfalch replied:
Good point - I'll try to remove all Swift code - also the #ifdef that I moved to Objective-c as you mentioned.

Kudo commented on packages/expo-background-task/ios/BackgroundTaskConsumer.swift:
```diff
@@ -0,0 +1,72 @@
+  func didBecomeReadyToExecute(withData data: [AnyHashable: Any]?) {
+    // Run on main thread. The task execution needs to be called on the main thread
+    // since it accesses the UIApplication's state
+    EXUtilities.performSynchronously {
```
```suggestion
    assert(Thread.isMainThread)
    EXUtilities.performSynchronously {
```

chrfalch replied:
I don't think we want to assert here - because we don't know where this will be called from (Task Manager calls this code from the iOS `BGTaskScheduler`'s thread) - what we want is to make sure the `self.task?.execute(withData: data, withError: nil)` line is called on the main thread - if this makes sense.

Kudo replied:
oh np! i just misled by the `// Run on main thread` comment

Kudo commented on packages/expo-background-task/ios/BackgroundTaskScheduler.swift:
```diff
@@ -0,0 +1,65 @@
+   Tries to schedule the worker task to run
+   */
+  @objc public static func tryScheduleWorker() throws {
+    print("BackgroundTaskScheduler: tryScheduleWorker")
```
```suggestion
```

Kudo commented on packages/expo-background-task/ios/BackgroundTaskScheduler.swift:
```diff
@@ -0,0 +1,65 @@
+   Cancels the worker task
+   */
+  public static func stopWorker() {
+    print("BackgroundTaskScheduler: stopWorker")
```
```suggestion
```

Kudo commented on packages/expo-background-task/ios/BackgroundTaskTester.h:
```diff
@@ -0,0 +1,18 @@
+/**
+ * This interface/implementation in Objective-C ensures that the triggerBackgroundTaskTest is
+ * only compiled in debug mode. There is some uncertainty about whether Swift's #if
+ * conditionals work the same way as in Objective-C.
```
yes, swift supports `#if DEBUG`. i had the same question before and found a proof.

chrfalch replied:
Awesome - good to know. I've implemented it in Objective-c for now :)

Kudo commented on packages/expo-background-task/src/BackgroundTask.ts:
```diff
@@ -0,0 +1,104 @@
+  }
+}
+
+// Export types
+export {
```
```suggestion
export type {
```

chrfalch replied:
Not sure if this works - often we import like `import * as BackgroundTask from ...` and then BackgroundTask.BackgroundTaskStatus wouldn't work.

Kudo reviewed (commented):
sorry late for the review. having some nit comments, not sure if makes sense to you.
also i think we have too many `Log.i` on android. maybe use `Log.d` or even remove unnecessary log before shipping.

chrfalch committed:
small adjustment in docs

chrfalch committed:
[codereview]: removed activity from debugging

chrfalch commented on packages/expo-background-task/android/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,10 @@
+
+  <application>
+    <activity
+      android:name=".BackgroundTaskDevActivity"
```
I've removed it - it was something from when I debugged this.

chrfalch commented on packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskConsumer.kt:
```diff
@@ -0,0 +1,116 @@
+const val DEFAULT_INTERVAL_MINUTES = 60L * 24L // Once every day
+private val TAG = BackgroundTaskConsumer::class.java.simpleName
+
+class BackgroundTaskConsumer(context: Context?, taskManagerUtils: TaskManagerUtilsInterface?) :
```
I think it will need to be public since it is used by the Task Manager through registration of tasks?

chrfalch committed:
[codereview] applied comments from code review from Kudo

chrfalch committed:
log.i -> log.d

chrfalch committed:
[coderview] reverted change

chrfalch committed:
[codereview] Made exceptions internal

chrfalch committed:
[codereview] Removed unnessecary constant

chrfalch committed:
[codereview] cleanup up declaration of TAG in module.

chrfalch committed:
[codereview] Changed to be swift-only

- Obj-c app-delegate -> swift AppDelegate
- Debug helper from obj-c -> swift
- Added runWithReason to EXTaskServiceInterface
- Removed obj-c annotations from swift code
- Changed scheduling to only be active when app is in the background
- Changed scheduling to use the minimal interval from JS code on iOS - which is supported through the `earliestBeginDate` field on the task request.

chrfalch committed:
[codereview] interface -> type

chrfalch committed:
[lint] swiftlint

amandeepmittal commented on docs/pages/versions/unversioned/sdk/background-task.mdx:
```diff
@@ -0,0 +1,208 @@
+
+## Background tasks
+
+A background task is a deferrable unit of work that is performed in the background, outside the lifecycle of your app. This is useful for tasks that need to be executed when the app is inactive, such as syncing data with a server, fetching new content, or even checking if there are any [Expo Updates](updates.mdx).
```
```suggestion
A background task is a deferrable unit of work that is performed in the background, outside your app's lifecycle. This is useful for tasks that need to be executed when the app is inactive, such as syncing data with a server, fetching new content, or even checking if there are any [`expo-updates`](updates.mdx).
```

- Let's use the Expo package name here instead of the API. Generally, users are more familiar with the package names.
- We can be concise at the end of the first sentence.

amandeepmittal commented on docs/pages/versions/unversioned/sdk/background-task.mdx:
```diff
@@ -0,0 +1,208 @@
+
+The native platform APIs will stop running background tasks if the user explicitly kills the app (they will be resumed when the user restarts the app) or if the app is uninstalled. If the device is rebooted or the app is killed by the system, the background tasks will be resumed.
+
+> **Info** The non-deterministic behavior of background tasks is intentional, aimed at optimizing battery life and enhancing user experience. If a user force stops an app, the system interprets this as a signal not to run background tasks for that app until the user restarts it. On iOS an app is force stopped if the user swipes up on the app in the app switcher. On Android this behaviour depends  ...
```
This is a bit of a large text to call out. Can it be the second paragraph of the same section?

chrfalch replied:
Yes, makes sense. I've refactored the whole section and simplified these sections a bit after some discussions with Brent as well. 

Also refactored the part of the doc about testing (will create a devtools plugin to ease testing).

amandeepmittal replied:
Thank you!

amandeepmittal commented on docs/pages/versions/unversioned/sdk/background-task.mdx:
```diff
@@ -0,0 +1,208 @@
+
+The native platform APIs will stop running background tasks if the user explicitly kills the app (they will be resumed when the user restarts the app) or if the app is uninstalled. If the device is rebooted or the app is killed by the system, the background tasks will be resumed.
+
+> **Info** The non-deterministic behavior of background tasks is intentional, aimed at optimizing battery life and enhancing user experience. If a user force stops an app, the system interprets this as a signal not to run background tasks for that app until the user restarts it. On iOS an app is force stopped if the user swipes up on the app in the app switcher. On Android this behaviour depends  ...
```
```suggestion
> **Info** The non-deterministic behavior of background tasks is intentional, aimed at optimizing battery life and enhancing user experience. If a user force stops an app, the system interprets this as a signal not to run background tasks for that app until the user restarts it.On Android, this behavior depends on the device vendor. On iOS, an app is force stopped if the user swipes up on the app in the app switcher.  Some vendors have implementations where the app is force stopped when the user removes the app from the recent apps list.
```

- Let's apply the platform order preference for Android and iOS.
- We prefer to use American English spellings in our docs [reference: Writing style guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#voice-and-tone).

amandeepmittal commented on docs/pages/versions/unversioned/sdk/background-task.mdx:
```diff
@@ -0,0 +1,208 @@
+
+### iOS&ensp;<PlatformTags platforms={['ios']} />
+
+The [`Background Tasks`](https://developer.apple.com/documentation/backgroundtasks) API on iOS is not available on simulators, it is only available when running on a physical device.
```
```suggestion
The [`Background Tasks`](https://developer.apple.com/documentation/backgroundtasks) API is unavailable on iOS simulators. It is only available when running on a physical device.
```

This can be concise.

amandeepmittal commented on docs/pages/versions/unversioned/sdk/background-task.mdx:
```diff
@@ -0,0 +1,208 @@
+
+To be able to run background tasks on iOS, you need to add the `processing` value to the `UIBackgroundModes` array in your app's **Info.plist** file. This is required for background fetch to work properly.
+
+**If you're using [CNG](/workflow/continuous-native-generation/)**, the required `UIBackgroundModes` configuration will be applied automatically by prebuild.
```
Can we make this part the first paragraph in this section and then the current first paragraph (`you need to add the `processing`) can follow with the code snippet so we are not repeating the information? Right now, we are repeating the part both in the first paragraph and in the collapsible about adding `processing` value manually.

amandeepmittal commented on docs/pages/versions/unversioned/sdk/background-task.mdx:
```diff
@@ -0,0 +1,208 @@
+
+## Usage
+
+Below is an example that demonstrates how to use `expo-background-task`.
```
```suggestion
Below is an example that demonstrates how to use `expo-background-task`:
```

amandeepmittal commented on docs/pages/versions/unversioned/sdk/background-task.mdx:
```diff
@@ -0,0 +1,208 @@
+
+## Multiple background tasks
+
+Since the Background Tasks API on iOS and the WorkManager API on Android limit the number of tasks that can be scheduled for a single app, Expo Background Task uses a single worker on both platforms. While you can define multiple JavaScript background tasks, they will all run through this single worker.
```
```suggestion
Since the WorkManager API on Android and the Background Tasks API on iOS limit the number of tasks that can be scheduled for a single app, Expo Background Task uses a single worker on both platforms. While you can define multiple JavaScript background tasks, they will all run through this single worker.
```

amandeepmittal commented on packages/expo-background-task/src/BackgroundTask.ts:
```diff
@@ -0,0 +1,104 @@
+ * Returns the status for the Background Task API. On web, it always returns `BackgroundTaskStatus.Restricted`,
+ * while on native platforms it returns `BackgroundTaskStatus.Available`. There is
+ *
+ * @returns A BackgroundTaskStatus enum value or null if not available.
```
```suggestion
 * @returns A [BackgroundTaskStatus](#backgroundtaskstatus) enum value or null if not available.
```

Let's link this to the Enum property and also add an inline highlight.

Simek replied:
```suggestion
 * @returns A BackgroundTaskStatus enum value or `null` if not available.
```

We should also format `null`.

amandeepmittal commented on packages/expo-background-task/src/BackgroundTask.types.ts:
```diff
@@ -0,0 +1,44 @@
+  /**
+   * Inexact interval in minutes between subsequent repeats of the background tasks. The final
+   * interval may differ from the specified one to minimize wakeups and battery usage.
+   * - Defaults to once every day on Android (The minimum interval is 15 minutes)
```
```suggestion
   * - Defaults to once every day on Android (the minimum interval is 15 minutes)
```

amandeepmittal commented on packages/expo-background-task/src/BackgroundTask.types.ts:
```diff
@@ -0,0 +1,44 @@
+   * Inexact interval in minutes between subsequent repeats of the background tasks. The final
+   * interval may differ from the specified one to minimize wakeups and battery usage.
+   * - Defaults to once every day on Android (The minimum interval is 15 minutes)
+   * - On iOS the system will determine the interval.
```
```suggestion
   * - On iOS, the system will determine the interval
```

For consistency, let's remove the period at the end of the sentence since this is list item.

amandeepmittal commented on packages/expo-background-task/src/BackgroundTask.types.ts:
```diff
@@ -0,0 +1,44 @@
+
+// @needsAudit
+/**
+ * Options for registering a background task
```
```suggestion
 * Options for registering a background task.
```

amandeepmittal commented on packages/expo-background-task/src/BackgroundTask.types.ts:
```diff
@@ -0,0 +1,44 @@
+// @needsAudit
+/**
+ * Availability status for background tasks
```
```suggestion
 * Availability status for background tasks.
```

amandeepmittal commented on packages/expo-background-task/src/BackgroundTask.types.ts:
```diff
@@ -0,0 +1,44 @@
+ */
+export enum BackgroundTaskStatus {
+  /**
+   * Background updates are unavailable
```
```suggestion
   * Background updates are unavailable.
```

amandeepmittal commented on packages/expo-background-task/src/BackgroundTask.types.ts:
```diff
@@ -0,0 +1,44 @@
+   */
+  Restricted = 1,
+  /**
+   * Background updates are available for the app
```
```suggestion
   * Background updates are available for the app.
```

amandeepmittal commented on packages/expo-background-task/src/BackgroundTask.types.ts:
```diff
@@ -0,0 +1,44 @@
+
+// @needsAudit
+/**
+ * Return value for background tasks
```
```suggestion
 * Return value for background tasks.
```

amandeepmittal commented on packages/expo-background-task/src/BackgroundTask.types.ts:
```diff
@@ -0,0 +1,44 @@
+ */
+export enum BackgroundTaskResult {
+  /**
+   * The task finished successfully
```
```suggestion
   * The task finished successfully.
```

amandeepmittal commented on packages/expo-background-task/src/BackgroundTask.types.ts:
```diff
@@ -0,0 +1,44 @@
+   */
+  Success = 1,
+  /**
+   * The task failed
```
```suggestion
   * The task failed.
```

chrfalch committed:
[codereview] Simplified kotlin code

- Removed some complex coroutine stuff
- Only start worker when app is moved to the background
- Clear worker when app is foregrounded again
- Cleaned up logging

chrfalch committed:
Update docs/pages/versions/unversioned/sdk/background-fetch.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

chrfalch committed:
Update docs/pages/versions/unversioned/sdk/background-task.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

chrfalch committed:
Update packages/expo-background-task/src/BackgroundTask.types.ts

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

chrfalch committed:
Update packages/expo-background-task/src/BackgroundTask.types.ts

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

chrfalch committed:
Update packages/expo-background-task/src/BackgroundTask.types.ts

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

chrfalch committed:
Update packages/expo-background-task/src/BackgroundTask.types.ts

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

chrfalch committed:
Update packages/expo-background-task/src/BackgroundTask.types.ts

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

chrfalch committed:
Update docs/pages/versions/unversioned/sdk/background-task.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

chrfalch committed:
Update docs/pages/versions/unversioned/sdk/background-task.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

chrfalch committed:
[codereview] updated titles/build for types

Simek commented on packages/expo-background-task/src/BackgroundTask.ts:
```diff
@@ -0,0 +1,104 @@
+// @needsAudit
+/**
+ * Returns the status for the Background Task API. On web, it always returns `BackgroundTaskStatus.Restricted`,
+ * while on native platforms it returns `BackgroundTaskStatus.Available`. There is
```
```suggestion
 * while on native platforms it returns `BackgroundTaskStatus.Available`.
```

Partially unfinished sentence.

chrfalch committed:
[codereview] simplified docs

- Simplified the sections about running/stopping
- Updated testing to include reference to upcoming dev tools plugin
- generated API docs

chrfalch committed:
Update packages/expo-background-task/src/BackgroundTask.ts

Co-authored-by: Bartosz Kaszubowski <gosimek@gmail.com>

chrfalch committed:
Update packages/expo-background-task/src/BackgroundTask.ts

Co-authored-by: Bartosz Kaszubowski <gosimek@gmail.com>

chrfalch committed:
[codereview] regenerated build files

chrfalch committed:
Merge branch 'main' into feature/expo-background-task-2

chrfalch committed:
[] Merge branch 'main' into feature/expo-background-task-2

chrfalch committed:
[X] Merge branch 'main' into feature/expo-background-task-2

chrfalch committed:
updated triggerTaskWorkerForTestingAsync to return true/false

To indicate to the caller (especially when debugging) that the call was successfull, the method has been updated to return true/false to indicate wether or not background tasks where run.

chrfalch committed:
swiftlint

lukmccall commented on apps/native-component-list/src/screens/BackgroundTaskScreen.tsx:
```diff
@@ -0,0 +1,148 @@
+    <View style={styles.screen}>
+      <View style={styles.textContainer}>
+        <Text>
+          Background Task Service:{' '}
```
```suggestion
          Background Task Service:
```

lukmccall commented on packages/expo-background-task/android/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,3 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest xmlns:android="http://schemas.android.com/apk/res/android">
+</manifest>
```
```suggestion
</manifest>

```

lukmccall commented on packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskScheduler.kt:
```diff
@@ -0,0 +1,231 @@
+        return false
+      }
+
+      val tasks = consumers.mapNotNull { consumer ->
+        val bgTaskConsumer = consumer as? BackgroundTaskConsumer ?: return@mapNotNull null
+
+        Log.d(TAG, "runTasks: executing tasks for consumer of type ${consumer.taskType()}")
+
+        val taskCompletion = CompletableDeferred<Unit>()
+
+        bgTaskConsumer.executeTask {
+          Log.d(TAG, "Task successfully finished")
+          taskCompletion.complete(Unit)
+        }
+
+        taskCompletion
+      }
+
+      // Await all tasks to complete
+      tasks.awaitAll()
+
+      return true
+    }
```
```suggestion
      val tasks = consumers
        .filterIsInstance<BackgroundTaskConsumer>()
        .map { bgTaskConsumer ->
          Log.d(TAG, "runTasks: executing tasks for consumer of type ${consumer.taskType()}")

          val taskCompletion = CompletableDeferred<Unit>()

          bgTaskConsumer.executeTask {
            Log.d(TAG, "Task successfully finished")
            taskCompletion.complete(Unit)
          }

          taskCompletion
        }

      // Await all tasks to complete
      tasks.awaitAll()

      return true
    }
```

lukmccall commented on packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskScheduler.kt:
```diff
@@ -0,0 +1,231 @@
+     * Helper function for calling functions returning an Operation
+     */
+    suspend fun Operation.await(): Boolean = withContext(Dispatchers.IO) {
+      try {
+        result.get()
+        true
+      } catch (e: Exception) {
+        throw e
+      }
```
```suggestion
      result.get()
      true
```
This try-catch seems to be unnecessary 🤔

chrfalch committed:
Update apps/native-component-list/src/screens/BackgroundTaskScreen.tsx

Co-authored-by: Łukasz Kosmaty <lukasz.kosmaty@swmansion.com>

chrfalch committed:
Update packages/expo-background-task/android/src/main/AndroidManifest.xml

Co-authored-by: Łukasz Kosmaty <lukasz.kosmaty@swmansion.com>

chrfalch committed:
removed triggerTaskWorkerForTestingAsync from docs

- Untill we know how we'd like the dev tools plugin to work I'm omitting the test method from documentation and public interface.

- Also correct a suggested rephrasing

chrfalch committed:
Removed test button

chrfalch committed:
Reverted removing parameter in bg screen

chrfalch committed:
Fixed how default interval was set in swift.

Now also defaults to 12 hours.

chrfalch committed:
Android: interval is default set to 12 hours now.

Removed difference between release and debug

chrfalch committed:
updated docs with correct info about interval

chrfalch committed:
applied last codereview comments

Kudo commented on packages/expo-background-task/android/src/main/java/expo/modules/backgroundtask/BackgroundTaskScheduler.kt:
```diff
@@ -0,0 +1,214 @@
+import java.time.Duration
+import java.util.concurrent.TimeUnit
+
+class BackgroundTaskScheduler {
+  companion object {
```
```suggestion
object BackgroundTaskScheduler {
```

if the whole class is in a companion object, we can just use `object`

Kudo commented on packages/expo-background-task/ios/BackgroundTaskConstants.swift:
```diff
@@ -0,0 +1,13 @@
+// Copyright 2024-present 650 Industries. All rights reserved.
+import Foundation
+
+public class BackgroundTaskConstants: NSObject {
```
```suggestion
public struct BackgroundTaskConstants {
```

can just use struct and not necessarily derive NSObject

Kudo commented on packages/expo-background-task/src/ExpoBackgroundTaskModule.ts:
```diff
@@ -0,0 +1,13 @@
+import { requireNativeModule } from 'expo-modules-core';
+import type { NativeModule } from 'expo-modules-core';
```
```suggestion
import { requireNativeModule, type NativeModule } from 'expo';
```

these are re-exported from the expo package

chrfalch committed:
codereview: updated with comments from Kudo

Thanks for the input @kudo :)
